src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        /// <summary>
+        /// Maintains a cache of the most recent variable states and manages
+        /// taking a snapshot of the NullableWalker in order to be as efficient
+        /// as possible when creating many snapshots.
+        /// </summary>
+        private sealed class Snapshotter
+        {
+            private readonly NullableWalker _walkerToSnapshot;
+            private ImmutableDictionary<VariableIdentifier, int> _lastVariableSlot = ImmutableDictionary<VariableIdentifier, int>.Empty;
+            private ImmutableArray<VariableIdentifier> _lastVariableBySlot = ImmutableArray<VariableIdentifier>.Empty;
+            private readonly PooledDictionary<BoundNode, Snapshot> _snapshotMap;
```

Comment:
>BoundNode [](start = 46, length = 9)

Since this is meant to serve speculative SemanticModel APIs and those use positions to identify the place to speculate at, it feels like it would be more natural to use position as a key in this map as well.